### ISSUE TITLE
New pattern to consider for testing Terraform code

### ISSUE THREAD

**Issue by brikis98** [gruntwork-io/terratest#223]
I've been contemplating a simple wrapper method we can create to simplify the typical pattern for testing Terraform code. We add a new method to the `terraform` package with a signature similar to the following:

```go
func TestModule(t *testing.T, options *terraform.Options, validate func(outputs *terraform.Outputs))
```

When you call this method, you pass it an `options` object with the path to the Terraform module to test and the variables to set for it, and `TestModule` does the following:

1. Use `test_structure.CopyTerraformFolderToTemp` to copy your code to a temp folder so that concurrent tests against the same Terraform code don't clash.
1. Run `terraform init`.
1. Run `terraform apply`, forwarding your vars along.
1. Call your `validate` function, passing it all the outputs from your module.
1. Call `terraform destroy` at the end of the test.
1. Delete the temp folder at the end of the test.
1. Note that the `apply` and `destroy` steps would be wrapped in `test_structure.RunTestStage` with well-known names, so you can skip one or the other or both while iterating in dev mode.

This reduces a typical Terraform test to:

```go
func MyTest(t *testing.T) {
  options := &terraform.Options {
    Path: "../my-module",
    Vars: map[string]interface{
      "foo": "bar"
    },
  }

  terraform.TestModule(t, options, func(outputs *terraform.Outputs) {
    publicIp := outputs.get("public_ip") as string
    http_helper.HttpGetWithRetry(t, publicIp)
  })
}
```

**Comment by yorinasub17** [comment]
Should this also call `test_structure.CopyTerraformFolderToTemp` by default? It makes debugging harder, but at the same time a lot of test isolation issues go away when we do this and I almost feel like this should be a standard pattern.

**Comment by brikis98** [comment]
Yes! I meant to include that but totally forgot. Thx for the reminder!

It might also automatically create some "stages" for you so you can run only `apply` or only `destroy` if you need to. I'll update the original description.

**Comment by brikis98** [comment]
BTW, now that I think about it, I wonder if we could avoid the need to copy the Terraform folder around if we: 

1. Add a `StateFileName` field to `terraform.Options`.
1. Set it to a unique name using the `UniqueId` method for each test.
1. Set `-state=<StateFileName>` whenever calling `terraform apply`, `terraform destroy`, etc.

That is, by ensuring each test gets its own state file, they can all run in the same folder without clashing.

**Comment by floresj** [comment]
This is pretty awesome. Is this something you all are thinking of including in the near future? As a little workaround, I'm experimenting with using the [TF_CLI_ARS](https://www.terraform.io/docs/commands/environment-variables.html#tf_cli_args-and-tf_cli_args_name)  environment variables to get that same behavior.


```go
uniqueID := random.UniqueId() 
tfStatePath := fmt.Sprintf("-state=%s.terraform.tfstate", uniqueID)

terraformOptions := &terraform.Options {
    EnvVars: map[string]string {
        "TF_CLI_ARGS_apply": tfStatePath,
        "TF_CLI_ARGS_destroy": tfStatePath,
    }
    // Other options...
}
```

**Comment by brikis98** [comment]
Would love to add this, but been busy on TF 0.12 upgrades. PR welcome in the meantime!

**Comment by nmiodice** [comment]
@brikis98 I've been working on testing automation that does pretty much exactly what you are describing. I'm interested to see if this can be contributed back to Terratest. I've supported both unit and integration testing patterns and have done a fair bit of orchestration to clean up any resources that are created. There are also some helpers to do basic sanity checking like
- The correct number of resources/outputs are created
- Assert values from plan/output for correctness, including a "subset" equality match where you can specify only what you care about asserting
- Hooks for custom validation functions and workspace overrides

This seems like exactly what you are talking about. Let me know how we can collaborate on this!

[Codebase](https://github.com/microsoft/cobalt/tree/master/test-harness/infratests)
[Tests example 1](https://github.com/microsoft/cobalt/tree/master/infra/templates/az-service-single-region/tests)
[Tests example 2](https://github.com/microsoft/cobalt/tree/master/infra/templates/az-isolated-service-single-region/tests)

**Comment by brikis98** [comment]
@nmiodice Oh, intriguing! From a quick glance, this looks very promising. A PR to add some of this to Terratest would be very welcome 👍 

As a first step, do you have a README—or could you create one—describing the approach? I'm getting a basic idea of it from the examples (i.e., validating `plan` output as a unit test, having a structured way to run `apply` and validate real running resources for integration tests), but would love a more formal write up so I can wrap my head around it.

**Comment by brikis98** [comment]
@nmiodice I just poked around the code examples some more and this looks very promising. Love the idea of how you structure unit and integration tests. PR to add this would be fantastic.

**Comment by james03160927** [comment]
Closing this issue due to 6 years of inactivity. Feel free to reopen if it still needs to be addressed.
